tests/ast/decoder.py

```diff
@@ -554,71 +557,49 @@ def decode_expr(self, expr: proto.Expr) -> Any:
-                return col_left != col_right
+                return self.binop(expr.neq, lambda lhs, rhs: lhs != rhs)
 
             case "gt":
-                col_left = self.decode_expr(expr.gt.lhs)
-                col_right = self.decode_expr(expr.gt.rhs)
-                return col_left > col_right
+                return self.binop(expr.gt, lambda lhs, rhs: lhs > rhs)
 
             case "lt":
-                col_left = self.decode_expr(expr.lt.lhs)
-                col_right = self.decode_expr(expr.lt.rhs)
-                return col_left < col_right
+                return self.binop(expr.lt, lambda lhs, rhs: lhs < rhs)
 
             case "geq":
-                col_left = self.decode_expr(expr.geq.lhs)
-                col_right = self.decode_expr(expr.geq.rhs)
-                return col_left >= col_right
+                return self.binop(expr.geq, lambda lhs, rhs: lhs >= rhs)
 
             case "leq":
-                col_left = self.decode_expr(expr.leq.lhs)
-                col_right = self.decode_expr(expr.leq.rhs)
-                return col_left <= col_right
+                return self.binop(expr.leq, lambda lhs, rhs: lhs <= rhs)
 
             case "sub":
-                col_left = self.decode_expr(expr.sub.lhs)
-                col_right = self.decode_expr(expr.sub.rhs)
-                return col_left - col_right
+                return self.binop(expr.sub, lambda lhs, rhs: lhs - rhs)
 
             case "mul":
-                col_left = self.decode_expr(expr.mul.lhs)
-                col_right = self.decode_expr(expr.mul.rhs)
-                return col_left * col_right
+                return self.binop(expr.mul, lambda lhs, rhs: lhs * rhs)
 
             case "div":
-                col_left = self.decode_expr(expr.div.lhs)
-                col_right = self.decode_expr(expr.div.rhs)
-                return col_left / col_right
+                return self.binop(expr.div, lambda lhs, rhs: lhs / rhs)
 
             case "mod":
-                col_left = self.decode_expr(expr.mod.lhs)
-                col_right = self.decode_expr(expr.mod.rhs)
-                return col_left % col_right
+                return self.binop(expr.mod, lambda lhs, rhs: lhs % rhs)
 
             case "pow":
-                col_left = self.decode_expr(expr.pow.lhs)
-                col_right = self.decode_expr(expr.pow.rhs)
-                return col_left**col_right
+                return self.binop(expr.pow, lambda lhs, rhs: lhs**rhs)
 
             case "and":
                 # "and" is reserved keyword in python - so have to use getattr here.
-                col_left = self.decode_expr(getattr(expr, "and").lhs)
-                col_right = self.decode_expr(getattr(expr, "and").rhs)
-                return col_left & col_right
+                left = self.decode_expr(getattr(expr, "and").lhs)
+                right = self.decode_expr(getattr(expr, "and").rhs)
+                return left & right
```

Comment:
why not:
```
return self.binop(getattr(expr, "and"), lambda ...)
```

Reply:
Also, I believe this is logical `and`, not `&`, which is represented by `bitand`

Reply:
raise TypeError(
            "Cannot convert a Column object into bool: please use '&' for 'and', '|' for 'or', "
            "'~' for 'not' if you're building DataFrame filter expressions. For example, use df.filter((col1 > 1) & (col2 > 2)) instead of df.filter(col1 > 1 and col2 > 2)."
        )
E       TypeError: Cannot convert a Column object into bool: please use '&' for 'and', '|' for 'or', '~' for 'not' if you're building DataFrame filter expre